### ISSUE TITLE
RecordEventsReadableSpan.getAttributes/toSpanData not properly synchronized

### ISSUE THREAD

**Issue by Oberon00** [open-telemetry/opentelemetry-java#699]
See: https://github.com/open-telemetry/opentelemetry-java/blob/ba05a6514f69c7c4605d637fc965c7145513b6de/sdk/src/main/java/io/opentelemetry/sdk/trace/RecordEventsReadableSpan.java#L288-L290

`RecordEventsReadableSpan.getAttributes` returns an unmodifieableMap, which is not a copy of but an (synchronized) reference to `attributes`. `toSpanData` calls `getAttributes` without further synchronization. Furthermore, `Span.attributes` is final, so the synchronized block in `getAttributes` could be elided.

What is more, the attributes are stored in a [`LinkedHashMap`](https://docs.oracle.com/javase/8/docs/api/java/util/LinkedHashMap.html) which is not thread safe and `toSpanData` also does not copy the attributes.

In the end `toSpanData` only really works if the span has been ended, otherwise it's race condition-prone.

**Comment by jkwatson** [comment]
Is the concern here that someone might add attributes after `toSpanData` has been called, but that the `SpanData` instance is intended to be a snapshot in time that is fully immutable?

**Comment by jkwatson** [comment]
I just verified that when we create the `SpanData` instance, we do make a copy of the attributes in the build method on the builder. So, the results on the `SpanData` are fully immutable, if this was the concern.

**Comment by Oberon00** [comment]
Indeed, I did miss the `new HashMap` here: https://github.com/open-telemetry/opentelemetry-java/blob/ba05a6514f69c7c4605d637fc965c7145513b6de/sdk/src/main/java/io/opentelemetry/sdk/trace/SpanData.java#L291

Then this is an invalid report, sorry for the noise.

**Comment by Oberon00** [comment]
I closed this too quickly. Even though the `new HashMap` is there in `SpanData.Builder.build()`, the call to `build()` in `Span.toSpanData()` is not synchronized, so `Span.setAttribute` could modify the map while `build` iterates over it.

**Comment by Oberon00** [comment]
See #706 / https://circleci.com/gh/open-telemetry/opentelemetry-java/1076: You'll get the following Exception if you try to call `setAttribute` concurrently with `toSpanData`: 

```
java.util.ConcurrentModificationException
        at java.util.LinkedHashMap$LinkedHashIterator.nextNode(LinkedHashMap.java:719)
        at java.util.LinkedHashMap$LinkedEntryIterator.next(LinkedHashMap.java:752)
        at java.util.LinkedHashMap$LinkedEntryIterator.next(LinkedHashMap.java:750)
        at java.util.Collections$UnmodifiableMap$UnmodifiableEntrySet$1.next(Collections.java:1663)
        at java.util.Collections$UnmodifiableMap$UnmodifiableEntrySet$1.next(Collections.java:1656)
        at java.util.HashMap.putMapEntries(HashMap.java:512)
        at java.util.HashMap.<init>(HashMap.java:490)
        at io.opentelemetry.sdk.trace.SpanData$Builder.build(SpanData.java:291)
        at io.opentelemetry.sdk.trace.RecordEventsReadableSpan.toSpanData(RecordEventsReadableSpan.java:179)
        at io.opentelemetry.sdk.trace.RecordEventsReadableSpanTest.testConcurrentModification(RecordEventsReadableSpanTest.java:620)
```

**Comment by Oberon00** [comment]
Fixed via #706.
